File: Code/Code/Matlab/basic_基础论文复现/preparing.m
% ======================================================================
% 本代码为论文《基于BP神经网络的自适应偏置比例导引》（刘畅等，兵工学报，2022）的复现尝试
% 引用时请注明原论文信息：
% 刘畅, 王江, 范世鹏, 等. 基于BP神经网络的自适应偏置比例导引[J]. 兵工学报, 2022, 43(11): 2798-2812.
% DOI: 10.12382/bgxb.2021.0594
% 
% ======================================================================
% 2025上半年“制导与控制”春季学期小组作业
% 组员：王宇轩、孙伟轩、常珈毓、罗嘉豪
% ======================================================================

%% 基于BP神经网络的自适应偏置比例导引算法数据生成模块
% 参考论文：《基于BP神经网络的自适应偏置比例导引》（刘畅等，兵工学报，2022）
% 功能：生成用于训练BP神经网络的样本数据集，包含四维输入参数与对应的偏置项b
% 关键逻辑：基于论文中的灵敏度分析结果分层采样，通过动力学仿真求解真实偏置项
% 对应论文第2.3节"样本建立"

% 参数定义
v = 300; % 导弹速度 (m/s)
% 生成参数范围（基于分层采样策略）
r0_part1 = 5000:100:7500;
r0_part2 = 7500+50:50:10000;
r0 = [r0_part1, r0_part2];
N_values = 2:1:4;
theta0 = 10:1:20;
qr = -70:0.5:-40;

% 生成四维参数空间样本
num_samples = 10000;
data = zeros(num_samples, 4);     % 输入：[r0, N, θ0, qr]
b_actual = zeros(num_samples, 1);

% 优化配置
options = optimset('TolX', 1e-6, 'Display', 'off', 'MaxIter', 100);

for i = 1:num_samples
    % 生成四维参数（均匀采样）
    data(i,1) = r0(randi(length(r0)));
    data(i,2) = N_values(randi(3));
    data(i,3) = theta0(randi(11));
    data(i,4) = qr(randi(length(qr)));
    
    % 运动学参数转换
    r0_val = data(i,1);
    N = data(i,2);
    theta0_val = deg2rad(data(i,3));
    qr_val = deg2rad(data(i,4));
    
    % 计算终端视线角（论文定义qr = qf - q0）
    q0 = 0; 
    qf = q0 + qr_val;
    % 终端弹道倾角θf = qf（论文假设条件）
    theta_f = qf;
    
    % 解析法，理论计算偏置项
    t_go = r0_val / v;
    b_guess = (N*q0 - theta0_val - (N-1)*theta_f) / t_go;
    
    % 优化求解实际b值，利用fzero优化器求解使终端角度误差为0的b值
    try
        b_actual(i) = fzero(@(b) compute_theta_error(b, N, v, r0_val,...
                          q0, theta0_val, theta_f), b_guess, options);
    catch
        b_actual(i) = b_guess;
        fprintf('优化失败样本%d，使用理论值b=%.4f\n', i, b_guess);
    end
end

% 保存数据集
inputs = data;               % [r0(m), N, θ0(deg), qr(deg)]
outputs = b_actual;          % 偏置项b(rad/s)
save('guidance_dataset.mat', 'inputs', 'outputs');

% 误差计算函数
function error = compute_theta_error(b, N, v, r0, q0, theta0, theta_f)
    % 弹目相对运动动力学模型
    function dydt = dynamics(~, y)
        r = y(1);    % 弹目距离
        q = y(2);    % 视线角
        theta = y(3);% 弹道倾角
        eta = theta - q;
        
        q_dot = -v * sin(eta) / r;
        theta_dot = N * q_dot + b;
        r_dot = -v * cos(eta);
        dydt = [r_dot; q_dot; theta_dot];
    end

    function [value, isterminal, direction] = stop_event(~, y)
        value = y(1) - 1;  % 当距离小于1米时停止
        isterminal = 1;    % 停止积分
        direction = -1;    % 检测下降沿
    end
    
    % 初始状态与仿真配置
    y0 = [r0; q0; theta0];
    tspan = [0, 1000];     % 最大仿真时间
    options = odeset('Events', @stop_event, 'RelTol', 1e-6);
    
    % 数值积分与误差计算
    try
        [~, y] = ode45(@dynamics, tspan, y0, options);
        if isempty(y)
            error = Inf;
        else
            theta_real = y(end, 3);
            error = theta_real - theta_f;
        end
    catch
        error = Inf;
    end
end

File: Code/Code/Python/preparing.py
import numpy as np
from scipy.optimize import root_scalar
from scipy.integrate import solve_ivp
import scipy.io as sio

def compute_theta_error(b, N, v, r0, q0, theta0, theta_f):
    """计算弹道倾角误差函数"""
    def dynamics(t, y):
        """导弹运动微分方程"""
        r, q, theta = y
        eta = theta - q
        
        q_dot = -v * np.sin(eta) / r
        theta_dot = N * q_dot + b
        r_dot = -v * np.cos(eta)
        return [r_dot, q_dot, theta_dot]
    
    def stop_event(t, y):
        """积分停止条件: 距离小于1米"""
        return y[0] - 1
    
    stop_event.terminal = True
    stop_event.direction = -1
    
    y0 = [r0, q0, theta0]
    t_span = (0, 100)  # 最大仿真时间
    
    try:
        sol = solve_ivp(dynamics, t_span, y0, events=stop_event, 
                        rtol=1e-6, atol=1e-8)
        if not sol.t_events[0].size:  # 没有触发停止事件
            return np.inf
        theta_real = sol.y[2, -1]  # 最终弹道倾角
        return theta_real - theta_f
    except:
        return np.inf

def main():
    # 参数定义
    v = 300  # 导弹速度 (m/s)
    
    # 生成参数范围
    r0_part1 = np.arange(5000, 7501, 100)    # 5000-7500m, 步长100m
    r0_part2 = np.arange(7550, 10001, 50)    # 7500-10000m, 步长50m
    r0 = np.concatenate((r0_part1, r0_part2))
    N_values = np.array([2, 3, 4])           # N ∈ {2,3,4}
    theta0 = np.arange(10, 21)               # θ0 ∈ [10,20]°
    qr = np.arange(-70, -39.5, 0.5)          # qr ∈ [-70,-40]°
    
    # 生成样本
    num_samples = 30000 # 样本个数
    data = np.zeros((num_samples, 4))        # 输入: [r0, N, θ0, qr]
    b_actual = np.zeros(num_samples)         # 输出: 偏置项b
    
    np.random.seed(42)  # 固定随机种子保证可重复性
    
    for i in range(num_samples):
        # 1. 生成四维参数(均匀采样)
        data[i, 0] = np.random.choice(r0)        # r0 ∈ [5,10] km
        data[i, 1] = np.random.choice(N_values)  # N ∈ {2,3,4}
        data[i, 2] = np.random.choice(theta0)    # θ0 ∈ [10,20]°
        data[i, 3] = np.random.choice(qr)       # qr ∈ [-70,-40]°
        
        # 2. 计算相关参数
        r0_val = data[i, 0]                     # 单位: m
        N = data[i, 1]
        theta0_val = np.deg2rad(data[i, 2])     # 单位: rad
        qr_val = np.deg2rad(data[i, 3])         # 单位: rad
        
        # 计算终端视线角
        q0 = 0                                  # 初始视线角为0°
        qf = q0 + qr_val                       # 终端视线角
        theta_f = qf                           # 终端弹道倾角θf = qf
        
        # 3. 理论计算偏置项(初始猜测值)
        t_go = r0_val / v                      # 飞行时间估计
        b_guess = (N*q0 - theta0_val - (N-1)*theta_f) / t_go
        
        # 4. 优化求解实际b值
        # 修改后的优化部分代码
        try:
            sol = root_scalar(
                lambda b: compute_theta_error(b, N, v, r0_val, q0, theta0_val, theta_f),
                x0=b_guess,
                x1=b_guess * 1.001,  # 提供第二个初始猜测点
                method='secant',       # 使用割线法
                xtol=1e-6,
                maxiter=100
            )
            b_actual[i] = sol.root
        except:
            b_actual[i] = b_guess
            print(f'优化失败样本{i}, 使用理论值b={b_guess:.4f}')
    
    # 保存数据集
    inputs = data                              # [r0(m), N, θ0(deg), qr(deg)]
    outputs = b_actual.reshape(-1, 1)          # 偏置项b(rad/s)
        
    # 如果需要PyTorch训练
    import torch
    torch.save({
        'inputs': torch.from_numpy(inputs).float(),
        'outputs': torch.from_numpy(outputs).float()
    }, 'guidance_dataset_python_3w.pt')

if __name__ == '__main__':
    main()
